adapter/stackdriver/stackdriver.go

```diff
@@ -0,0 +1,312 @@
+}
+
+func newFactory(createClient createClientFunc) *factory {
+	return &factory{DefaultBuilder: adapter.NewDefaultBuilder(adapterName, adapterDesc, &config.Params{}), createClient: createClient}
+}
+
+func (f *factory) Close() error {
+	f.t.Stop()
+	return nil
+}
+
+func createClient(cfg *config.Params) (*monitoring.MetricClient, error) {
+	return monitoring.NewMetricClient(context.Background(), toOpts(cfg)...)
+}
+
+// We keep this function separate from createClient to enable easy testing
+func toOpts(cfg *config.Params) (opts []gapiopts.ClientOption) {
+	switch cfg.Creds.(type) {
+	case *config.Params_ApiKey:
+		opts = append(opts, gapiopts.WithAPIKey(cfg.GetApiKey()))
+	case *config.Params_ServiceAccountPath:
+		opts = append(opts, gapiopts.WithServiceAccountFile(cfg.GetServiceAccountPath()))
+	case *config.Params_AppCredentials:
+		// When using default app credentials the SDK handles everything for us.
+	}
+	if cfg.Endpoint != "" {
+		opts = append(opts, gapiopts.WithEndpoint(cfg.Endpoint))
+	}
+	return
+}
+
+// NewMetricsAspect provides an implementation for adapter.MetricsBuilder.
+func (f *factory) NewMetricsAspect(env adapter.Env, c adapter.Config, metrics map[string]*adapter.MetricDefinition) (adapter.MetricsAspect, error) {
+	cfg := c.(*config.Params)
+	types := make(map[string]sdinfo, len(metrics))
+	for name, def := range metrics {
+		info, found := cfg.MetricInfo[name]
+		if !found {
+			env.Logger().Warningf("No stackdriver info found for metric %s, skipping it", name)
+			continue
+		}
+		// TODO: do we want to make sure that the definition conforms to stackdrvier requirements? Really that needs to happen during config validation
+		types[name] = sdinfo{
+			ttype: metricType(def.Name),
+			kind:  info.Kind,
+			value: info.Value,
+		}
+	}
+
+	// Per the documentation on config.proto, if push_interval is zero we'll default to a 5s push interval
+	if cfg.PushInterval == time.Duration(0) {
+		cfg.PushInterval = 1 * time.Second
+	}
+	f.once.Do(func() {
+		f.t = time.NewTicker(cfg.PushInterval)
+	})
+
+	var err error
+	var client *monitoring.MetricClient
+	// TODO: in theory this client could live in the factory and be shared amongst many adapter instances
+	if client, err = f.createClient(cfg); err != nil {
+		return nil, env.Logger().Errorf("Failed to construct stackdriver client: %v", err)
+	}
+
+	s := &sd{
+		env:         env,
+		projectID:   cfg.ProjectId,
+		client:      client,
+		pushMetrics: client.CreateTimeSeries,
+		metricInfo:  types,
+		m:           sync.Mutex{},
+	}
+	s.startTicker(env, f.t)
+	return s, nil
+}
+
+// Extracted from NewMetricsAspect for testing - it shouldn't be possible to get an Env any other time.
+func (s *sd) startTicker(env adapter.Env, ticker *time.Ticker) {
+	env.ScheduleDaemon(func() {
+		for range ticker.C {
+			s.pushData()
+		}
+	})
+}
+
+func (s *sd) Record(vals []adapter.Value) error {
+	s.env.Logger().Infof("stackdriver.Record called with %d vals", len(vals))
```

Comment:
It might be nice to maybe make this Verbosity-sensitive (i'm thinking we could see a million lines of stackdriver.Record).